checker/jdk/index/src/java/nio/CharBuffer.java

```diff
@@ -0,0 +1,1523 @@
+    public int compareTo(CharBuffer that) {
+        int n = this.position() + Math.min(this.remaining(), that.remaining());
+        for (int i = this.position(), j = that.position(); i < n; i++, j++) {
+            int cmp = compare(this.get(i), that.get(j)); // #1
+            if (cmp != 0)
+                return cmp;
+        }
+        return this.remaining() - that.remaining();
+    }
+
+    private static int compare(char x, char y) {
+
+
+
+
+
+
+        return Character.compare(x, y);
+
+    }
+
+    // -- Other char stuff --
+
+
+
+    /**
+     * Returns a string containing the characters in this buffer.
+     *
+     * <p> The first character of the resulting string will be the character at
+     * this buffer's position, while the last character will be the character
+     * at index <tt>limit()</tt>&nbsp;-&nbsp;1.  Invoking this method does not
+     * change the buffer's position. </p>
+     *
+     * @return  The specified string
+     */
+    @SuppressWarnings("index:return.type.incompatible") // As in documentation, this function returns a string containing the characters in the buffer
+    public @SameLen("this") String toString() {
+        return toString(position(), limit());
+    }
+
+    abstract String toString(int start, int end);       // package-private
+
+
+    // --- Methods to support CharSequence ---
+
+    /**
+     * Returns the length of this character buffer.
+     *
+     * <p> When viewed as a character sequence, the length of a character
+     * buffer is simply the number of characters between the position
+     * (inclusive) and the limit (exclusive); that is, it is equivalent to
+     * <tt>remaining()</tt>. </p>
+     *
+     * @return  The length of this character buffer
+     */
+    @SuppressWarnings("index:return.type.incompatible") // remaining is equivalent to the length of the character buffer as explained in the documentation
+    public final @NonNegative @LTEqLengthOf("this") int length() {
+        return remaining();
+    }
+
+    /**
+     * Reads the character at the given index relative to the current
+     * position.
+     *
+     * @param  index
+     *         The index of the character to be read, relative to the position;
+     *         must be non-negative and smaller than <tt>remaining()</tt>
+     *
+     * @return  The character at index
+     *          <tt>position()&nbsp;+&nbsp;index</tt>
+     *
+     * @throws  IndexOutOfBoundsException
+     *          If the preconditions on <tt>index</tt> do not hold
+     */
+    @SuppressWarnings({"index:override.param.invalid", "index:argument.type.incompatible"}) /*
+    index has to be @LTLengthOf("this.remaining()") as written in the documentation.
+    #1: index being @LTLengthOf("this.remaining()") => position() + checkIndex(index, 1) is @LTLengthOf("this.limit")
+    */
+    public final char charAt(@NonNegative @LTLengthOf("this.remaining()") int index) {
```

Comment:
I would do either `@LTLenghtOf("this")` or `@LessThan("this.remaining()")`, preferring the former. `@LTLengthOf("remaining()")` doesn't make sense, since `remaining` returns an `int`, not an array

Reply:
According to the documentation, it should be `@LessThan("this.remaining()")`. As the documentation says `The index of the character to be read, relative to the position must be non-negative and smaller than <tt>remaining()</tt>`.
This also makes sense because then `position() + checkIndex(index,1)` will also be `@LessThan("this.limit()")` as is required by the `get()` documentation.
The problem is with the `override.param.invalid` error, because the `charAt()` function in `CharSequence.java` cannot be annotated in terms of `limit()` as this class does not extend `Buffer.java`